### PR TITLE
Binary Classification Evaluator

### PR DESCRIPTION
**Pull request by y27choi**
### Linked issue(s):
Fixes KOL-2704

### What change does this PR introduce and why?
* Refactors multiclass classification evaluator
* Modified the workflow schema 

[dogs-vs-cats binary classification (threshold=0.5)](https://trunk.kolena.io/yoohee/results?modelIds=684&workflow=Image+Classification&testSuiteId=1370&evaluators=N4IgjAzAbAHCBcpZ3mADATgCwF8A04EA7GgkjFguhlAbAKxWa0ixRM045A)
[dogs-vs-cats multiclass classification](https://trunk.kolena.io/yoohee/results?modelIds=686&workflow=Image+Classification&testSuiteId=1370&evaluators=N4IgjAzAbAHCBcpZ3mADATgCwF8A04EA7GgkjFguhlAbAKxWa0ixRM045A)

Based on the # of prediction labels seeded as part of `Inference` determines which evaluator to run.  If only positive label's prediction is used, then it will be treated as a binary classification.  If both labels' predictions are added, then it will be 2 class multiclass evaluation. 

How are multiclass and binary evaluator different?:
* Per-class evaluation for multiclass (for both "dog" and "cat" class metrics are available) (e.g., TP for class "cat" is for correctly predicted "cat" class) — binary shows the same information but in different wordings (e.g., TP is for positive label which in this case is "dog" and TN is for negative label which in this case is "cat") 
* ROC curve on positive label only for binary vs on both labels for multiclass
* Score distribution (for binary, only the positive prediction scores are used so it ranges from 0 to 1; whereas, for multiclass, maximum confidence score label is used so it will range from 0.5 - 1.
* Classification label annotation: for binary, it will only show the prediction on the positive label and for multiclass, the label which has the "maximum" confidence score will be shown. 

Screen Shot 2023-08-03 at 9 58 53 AM Screen Shot 2023-08-03 at 9 59 02 AM

* TestSampleMetrics: for binary, there are more fields like `is_TP`, `is_FN` and etc., whereas for multiclass, only `is_correct` and predicted label is shown.

Screen Shot 2023-08-03 at 10 01 08 AM
Screen Shot 2023-08-03 at 10 01 20 AM

For binary classification cases where performance on positive label can be clearly defined (e.g., "is_live" or "is_dog") then using binary classification evaluator is a better experience.  For cases where both label are equally important (e.g., "dog" vs "cat") where a positive label cannot be specified, using multiclass evaluator is preferred.

Proposal on UX improvement:
Once we make this into a pre-built, we should consider changing the color of `TestSampleMetrics.classification` label to be green (if `is_correct=True`) or red (otherwise).

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

### CODE REVIEW

**Comment by y27choi** [comment]
will have a separate PR for seeding scripts and doc updates

File: examples/classification/classification/workflow.py [hunk 1]
```diff
@@ -65,24 +78,83 @@ class Inference(BaseInference):
...
+    """
+
+    missing_inference: bool = False
+
+    def __post_init__(self):
+        object.__setattr__(self, "missing_inference", self.classification is None)
+
+
 @dataclass(frozen=True)
 class TestSampleMetrics(MetricsTestSample):
```
Comment (ly-kolena) [inline examples/classification/classification/workflow.py:136]: Took me some time to figure out what's the difference between `TestSampleMetricsSingleClass` and `TestSampleMetrics` (hopefully I'm not wrong). They have the same class description. Is it intentional to have them the same? Would it be helpful to describe which class to use for which problem?

**Review by ly-kolena — APPROVED** [review]

File: examples/classification/classification/workflow.py (hunk 1 above)
Comment (y27choi) [reply examples/classification/classification/workflow.py:136]: Thanks for catching it.  No they are not supposed to be the same. I will update it to be more descriptive.

File: examples/classification/classification/evaluator_binary.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        plots.append(
+            compute_confusion_matrix(
+                [gt.classification.label for gt in ground_truths],
+                [positive_label if metric.is_TP or metric.is_FP else negative_label for metric in metrics],
+                labels=[positive_label, negative_label],
+            ),
+        )
+        plots = list(filter(lambda plot: plot is not None, plots))
+
+        return plots
```
Comment (gordonhart) [inline examples/classification/classification/evaluator_binary.py:124]: Should we add an Accuracy vs. T plot? On the [linked binary classification example](https://trunk.kolena.io/yoohee/results?modelIds=684&workflow=Image+Classification&testSuiteId=1370&evaluators=N4IgjAzAbAHCBcpZ3mADATgCwF8A04EA7GgkjFguhlAbAKxWa0ixRM045A), I find myself wanting to know the optimal threshold, which isn't quite conveyed in any of the existing plots.

**Review by gordonhart — APPROVED** [review]

File: examples/classification/classification/evaluator_binary.py (hunk 2 above)
Comment (y27choi) [reply examples/classification/classification/evaluator_binary.py:124]: That's a great idea. I will add a threshold plot (with accuracy, precision, recall, f1-score vs. T).

**Pull request merged**